File: linked-lists/reverse_linked_list.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

typedef struct ListNode {
    int value;
    struct ListNode *next;
} ListNode;

ListNode *reverse_list(ListNode *head) {
    ListNode *previous = NULL;
    while (head != NULL) {
        ListNode *next = head->next;
        head->next = previous;
        previous = head;
        head = next;
    }
    return previous;
}

int main(void) {
    ListNode third = {3, NULL};
    ListNode second = {2, &third};
    ListNode first = {1, &second};
    ListNode *reversed = reverse_list(&first);
    assert(reversed->value == 3 && reversed->next->value == 2 && reversed->next->next->value == 1);
    assert(reversed->next->next->next == NULL);

    assert(reverse_list(NULL) == NULL);
    puts("Reverse Linked List: 2 tests passed");
    return 0;
}

File: linked-lists/reverse_linked_list.py
"""LeetCode 206: Reverse Linked List."""
import unittest


class ListNode:
    def __init__(self, value=0, next_node=None):
        self.value = value
        self.next = next_node


def reverse_list(head):
    previous = None
    current = head
    while current:
        next_node = current.next
        current.next = previous
        previous, current = current, next_node
    return previous


def values_from(head):
    values = []
    while head:
        values.append(head.value)
        head = head.next
    return values


class ReverseLinkedListTests(unittest.TestCase):
    def test_typical_case(self):
        head = ListNode(1, ListNode(2, ListNode(3)))
        self.assertEqual(values_from(reverse_list(head)), [3, 2, 1])

    def test_edge_case_empty_list(self):
        self.assertIsNone(reverse_list(None))


if __name__ == "__main__":
    unittest.main()
